### ISSUE TITLE
Rule Request: Deprecation Count Rule

### ISSUE THREAD

**Issue by devandanger** [realm/SwiftLint#2444]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### New rule request
Deprecation Count rule would run like normal accept it would need to globally keep track of the number of violating deprecations.  This would be useful for general curation of the application.  If you have several deprecated custom class/struct you could prevent further re-usage of them.  

1. Why should this rule be added? Share links to existing discussion about what
   the community thinks about this.

2. Provide several examples of what _would_ and _wouldn't_ trigger violations.
3. Should the rule be configurable, if so what parameters should be configurable? Configurable with counts allowed for certain usages of a deprecated class/struct.
4. Should the rule be opt-in or enabled by default? Why?
opt-in.  Its usage it more similar to the custom_rule.
   See [README.md](../README.md#opt-in-rules) for guidelines on when to mark a rule as opt-in.

**Comment by marcelofabri** [comment]
Closing this because the template is not filled completely. Feel free to reopen this once we have more details.

You can also [mark declarations as deprecated with Swift attributes](https://www.hackingwithswift.com/example-code/language/how-to-use-available-to-deprecate-old-apis) which is probably a better tool for the job.

**Comment by devandanger** [comment]
@marcelofabri can you reopen this issue?... I don't have access to.

**Comment by devandanger** [comment]
Also regarding.

> You can also [mark declarations as deprecated with Swift attributes](https://www.hackingwithswift.com/example-code/language/how-to-use-available-to-deprecate-old-apis) which is probably a better tool for the job.

Looking into that approach, only problem is we have warnings/deprecations turned on 'as Errors'.

**Comment by marcelofabri** [comment]
It still doesn't make much sense to me to add this rule. One thing that bothers me is that which violations will be triggered is not deterministic: it depends on the order of the files that were checked.

It's also problematic because it introduce shared state (a singleton) between different files being checked - and we do it in parallel, so it can also create multithread issues.

If using `@availability` is not an option, I think it'd make more sense to use a custom rule and mark existing code with `// swiftlint:disable`.

**Comment by devandanger** [comment]
> It still doesn't make much sense to me to add this rule. One thing that bothers me is that which violations will be triggered is not deterministic: it depends on the order of the files that were checked.

Fair point, based on this implementation it wouldn't be perfect on which would trigger in the end.  It would still prevent the most recent one from being added which would do enough *i think* from deterring developers from doing the wrong thing.

> It's also problematic because it introduce shared state (a singleton) between different files being checked - and we do it in parallel, so it can also create multithread issues.

I can see this adding in a category of problems to the mix, but maybe there is a more generalized shared state that we could integrate into the library to add other features as well into the mix as well.  Understandably, the current singleton approach is crude; it was intended to be just a starting point.

> If using `@availability` is not an option, I think it'd make more sense to use a custom rule and mark existing code with `// swiftlint:disable`.

I started out with this approach using `// swiftlint` and just thought there was a better way, so started hacking at it this way.

If you feel strongly about this and don't think there's anything to work through with this that's fine I understand, there were alternative approaches I considered outside of this library.  I just felt like this was the best option to start pursuing.

**Comment by devandanger** [comment]
@marcelofabri any other thoughts on this?... I had a couple ideas regarding a different type of state to get around the shared state(singleton) but would make the changes more intensive.

**Comment by Dschee** [comment]
I must admit, I don't understand this rule request. What's the point of only showing warnings if a rule is broken at least a given number of times? If `EmbeddedViewController` shouldn't be used anymore, why is it ok to use it once but not twice? It doesn't make much sense to me.

Also as @marcelofabri noted, if you have a function or type that you want to mark not to be used anymore, you should simply use [Swift deprecation](https://www.hackingwithswift.com/example-code/language/how-to-use-available-to-deprecate-old-apis). Could you provide more info so I can understand it?

**Comment by devandanger** [comment]
So we have turned on warnings as errors.  Thats the crux of our problem honestly.

As iOS progresses we've come up with better implementations in isolated parts of the application.  These are older areas sometimes incurring low foot traffic and also low frequency of changes.  So there isn't always to compelling argument to go back and refactor them right away.  Classical technical debt discussion.  

We do want to discourage continuing older/less-than practices going forward.  We document this via other documentation/tech meetings/etc, however I wanted to see if there was an opportunity to use something like this in order to self-document the codebase.

I appreciate you two taking the time to look at this though (seriously thank you), but it doesn't sound like there's any interest in adding this into the Swiftlint though.  I'll close the related PR.

**Comment by Dschee** [comment]
I think if you have some older code parts in your project which might not adhere to the Swift best practices of the newer parts of your project, you could either completely exclude them from SwiftLint via the `exclude` configuration option or simply turn off SwiftLint (entirely via `all` or just the triggering rules) within the affected files via `// swiftlint:disable rule_name other_rule_name`. That's what the disabling commands are here for.

**Comment by devandanger** [comment]
Right that approach does work and it's what we are using right now.  Some of these parts that don't adhere are used in switch statements so this approach had us turning on and off this swiftlint between relevant case statements.  It wasn't too bad just didn't make the code look great in the end.

**Comment by Dschee** [comment]
I'm not sure what you code looks like, but with the different variants supporting `disable`/`enable` to surround a code partial or `disable:this` and `disable:next` to disble for specific lines only, there's quite some flexibility. You could also turn the rule off for the entirely file by adding a `disable` at the beginning of the file and noe `enable` at all. So you can play around with these options but I understand that none of these really looks perfect. But you'll need to make some compromise, whichever it is.
